### ISSUE TITLE
7.0.4 model with serialized field  hash_with_indifferent _access cannot be changed

### ISSUE THREAD

**Issue by salex** [rails/rails#46072]
### Steps to reproduce
I have several applications that used in a model

```ruby
serialize :settings, ActiveSupport::HashWithIndifferentAccess
```

That broke in 7.0.3 but a band-aid was provided:

```
  config.after_initialize do
    # Psych Allow YAML Classes
    config.active_record.yaml_column_permitted_classes = [Symbol, Hash, Array, Date, Range, ActiveSupport::HashWithIndifferentAccess, ActiveModel::Attribute.const_get(:FromDatabase), User, Time]
  end
```

I had converted one of the applications in 7.0.3 to use `serialize setting, JSON` and all worked fine. If this case it was a model that saved an object that didn't change unless any record in the model changed ` think accounts in an accounting application that rarely get changed, it just stores an object that has a  overview of the accounts, tree structure and all.`

I updated the `pg` gem in another application that had the serialization and it updated Rails to 7.0.4, I then got the serialization error that you can't do that:

```
/Users/salex/.rbenv/versions/3.0.1/lib/ruby/3.0.0/psych/class_loader.rb:99:in `find': 
Tried to load unspecified class: ActiveSupport::HashWithIndifferentAccess (Psych::DisallowedClass)
```

I tried to convert the serialized field to JSON , setting the field to nil and saving the model (it rebuilds if blank).  Can't do that.  That field appears to be forever frozen.

```
changing serialize in the model to JSON
getting a record
changing to value to nil
save the record
spits out the YAML that was originally saved, but you can't replace it!
```

### Expected behavior
I should be able to correct bad behavior

### System configuration
**Rails version**:
7.0.4
**Ruby version**:
3.1

Guess I could create a migration to delete the settings field and then another one to recreate it, but that seems like a hack,

**Comment by rafaelfranca** [comment]
In the application that you have the error, do you have `ActiveSupport::HashWithIndifferentAccess` inside `yaml_column_permitted_classes`?

**Comment by salex** [comment]
> In the application that you have the error, do you have `ActiveSupport::HashWithIndifferentAccess` inside `yaml_column_permitted_classes`?

Yes I did in `class Application < Rails::Application` 

```ruby
    config.after_initialize do
      # Psych Allow YAML Classes (found somewhere when error first occurred in 7.0.3)
      config.active_record.yaml_column_permitted_classes = [Symbol, Hash, Array, Date, Range, 
        ActiveSupport::HashWithIndifferentAccess, ActiveModel::Attribute.const_get(:FromDatabase), User, Time]
    end
```

That worked in 7.0.3, Not in 7.04. In my Book model I had:

```ruby
  serialize :settings, HashWithIndifferentAccess
```

I'm actually refactoring/rewriting the app and it's almost complete. In the rewrite I serialize with JSON and it works fine.

I figured the band-aid was not good enough and the error appears if you load a record:

```ruby
irb(main):001:0> b = Book.find 1
  Book Load (0.6ms)  SELECT "books".* FROM "books" WHERE "books"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
(Object doesn't support #inspect)
```

That seems to lock the record. I tried to set the setting attribute to nil and save the record and get:

```ruby
irb(main):003:0> b.settings = nil
=> nil
irb(main):004:0> b.save
/Users/salex/.rbenv/versions/3.0.1/lib/ruby/3.0.0/psych/class_loader.rb:99:in `find': Tried to load unspecified class: ActiveSupport::HashWithIndifferentAccess (Psych::DisallowedClass)
irb(main):005:0> 
```

I can probably comment out the serialize action, set field to nil and save, then serialize as plain Hash. My problem is I have used serialize in a couple other apps, but the data that can't be regenerated.  I'll have to convert my 7.0.3 apps to JSON serialization  before they are upgraded.

Is `config.active_record.yaml_column_permitted_classes` supposed to work in 7.0.4? If so, do I have something wrong in my method?

**Comment by natematykiewicz** [comment]
The problem is the `after_initialize`.

`yaml_column_permitted_classes` is a class-level attr_accessor on `ActiveRecord`.

https://github.com/rails/rails/blob/6ffa34dcf7faf891920119d4fc9599d9315e06bb/activerecord/lib/active_record.rb#L376-L377

```ruby
ActiveRecord.yaml_column_permitted_classes
=> [Symbol]
``` 

In an `after_initialize` hook, Rails goes through `Rails.application.config.active_record` and copies those settings over to `ActiveRecord`'s attr_accessors.
https://github.com/rails/rails/blob/6ffa34dcf7faf891920119d4fc9599d9315e06bb/activerecord/lib/active_record/railtie.rb#L223-L234

So you're setting the `Rails.application.config` too late. The settings have already been copied from there over to their corresponding libraries. It could make sense to freeze the `Rails.application.config` hashes at this point so that you got an error about modifying a frozen object.

Instead of:
```ruby
    config.after_initialize do
      # Psych Allow YAML Classes (found somewhere when error first occurred in 7.0.3)
      config.active_record.yaml_column_permitted_classes = [Symbol, Hash, Array, Date, Range, 
        ActiveSupport::HashWithIndifferentAccess, ActiveModel::Attribute.const_get(:FromDatabase), User, Time]
    end
```

Do:
```ruby
    # Psych Allow YAML Classes (found somewhere when error first occurred in 7.0.3)
    config.active_record.yaml_column_permitted_classes = [Symbol, Hash, Array, Date, Range, 
      ActiveSupport::HashWithIndifferentAccess, ActiveModel::Attribute.const_get(:FromDatabase), User, Time]
```

**Comment by salex** [comment]
[natematykiewicz](https://github.com/natematykiewicz) Thank you for the solution and the explanation. I'm just a hobbyist developer some stuff will go over my head.

As my comment states `# Psych Allow YAML Classes (found somewhere when error first occurred in 7.0.3)`, I found this somewhere and it worked in 7.0.3.x. I seem to remember something, somewhere in the 7.0.4 release comments that serializing hash date with indifferent access was a waste of time and it was removed or something.

Thanks again
